text/0000-pi-types.md

```diff
@@ -0,0 +1,644 @@
+```
+
+## `where` clauses
+
+Often, it is wanted to have some statically checked clause satisfied by the
+constant parameters (e.g., for the sake of compile-time bound checking). To
+archive this, in a reasonable manner, we use const exprs, returning a boolean.
+
+We allow such constexprs in `where` clauses of functions. Whenever the
+function is invoked given constant parameters `<a, b...>`, the compiler
+evaluates this expression, and if it returns `false`, an aborting error is
+invoked.
+
+To sum up, the check happens when typechecking the function calls (that is,
+checking if the parameters satisfy the trait bounds, and so on). The caller's
+bounds must imply the invoked functions' bounds:
+
+### Transitivity of bounds.
+
+We require a bound of a function to imply the bounds of the functions it calls,
+through a simple reductive, unification algorithm. In particular, this means
+that a statement is reduced by some specified rules (see below), that ensures
+termination. A statement implies another statement, if the set of statements it
+reduces to is a superset of the other statement's reduction set.
+
+The compiler would enforce that if `f` calls `g`, `unify(bound(g))	⊆
+unify(bound(f))` (by structural equality):
+
+    ExpandBooleanAnd:
+      P && Q
+      ──────
+      P
+      Q
+
+This simply means that `a && b` means `a` and `b`.
+
+    SubstituteEquality:
+      P(a)
+      a = b
+      ─────
+      P(b)
+
+This is an important inference rule, when doing unification. This means that
+you can substitute all `a` for all free `b`s, if `a = b`.
+
+    DoubleNegation:
+      ¬¬x
+      ───
+      x
+
+The last rule for now is simply stating that double negation is identity, that
+is, `!!a` means that `a` is true.
+
+These rules are "eliminatory" (recursing downwards the tree and decreasing the
+structure), and thus it is possible to check, in this language, that `a ⇒ b`
+relatively quickly (`O(n)`). For a proof of see the section below.
+
+More rules can be added in the future. It is however important to preserve the
+"sequential property" (that is, each step is a reduction, not an expansion),
+allowing one to check the implication in linear time.
+
+This is done under type unification. Thus, we only need to check the bounds at
+the top level.
+
+#### Decidability of this rule set
+
+One can show this by considering each case:
+
+1. `ExpandBooleanAnd` eliminates `{P && Q} ⊢ {P, Q}`. The right hand side's
+   depth is `max(dep(P), dep(Q))`, which is smaller than the original,
+   `max(dep(P), dep(Q)) + 1`
+2. `SubstituteEquality` eliminates `{a = b, P} ⊢ {P[b/a]}`, which is an
+   elimination, since `dep(P) + 1 > dep(P[b/a]) = dep(P)`.
+3. `DoubleNegation` eliminates `{¬¬x} ⊢ {x}`, which is an elimination, since
+   `dep(x) + 2 > dep(x)`.
+
+#### (Optional extension:) "Exit-point" identities
+
+These are simply identities which always holds. Whenever the compiler reach one
+of these in the `where` clause unfolding, it returns "True":
```

Comment:
reach => reaches

something weird with how unfolding is used, "in the unfolding `where` clause" maybe?